### PR TITLE
[C#] Add Pattern Matching Concept Exercise

### PR DESCRIPTION
**Pull request by MarkusReynolds1989**
Added my idea for the concept-exercise on pattern-matching.

### CODE REVIEW

**Comment by MarkusReynolds1989** [comment]
> Cool stuff @MarkusReynolds1989! Thanks a lot for your contribution so far.
> 
> Please don't get discouraged by the number of comments I've added. They're mostly very small nits that you should be able to fix just by accepting my changes in the GitHub UI.
> 
> I'm just starting out with v3, so @ErikSchierboom will be able give better feedback on how this fits into the overall picture of what we're looking for in terms of concept exercises.
> 
> The one thing that we haven't discussed yet that this PR highlights, is we don't really have a "story" to tell for this exercise like we do for the exercises with the log lines. I thought the idea is that we wanted to have stories around the exercises to make them feel less like you're writing code that you wouldn't use in the real world.

Great comments here and I will make sure I get these corrected. As for a story to tell, that is interesting, pattern matching and switches to me fall more into code glue/ general functionality. It's not really about one particular use case more so that you need it in most code you write.

**Comment by ErikSchierboom** [comment]
@MarkusReynolds1989 Thanks a ton for working on this! That said, there is one step that has been skipped here, and that is having the initial exercise design discussion. For new Concept Exercises, we want to have an upfront discussion. This both serves as a way for other maintainers to keep up to date with what's happening and give their insights on the exercise, besides it also serving as vital documentation. As there is so much going on with Concept Exercises, having an upfront discussion also helps prevent work having to be re-done.

@MarkusReynolds1989 Would you mind creating the corresponding GitHub issue so we can discuss the exercise there? There already is an example issue https://github.com/exercism/v3/issues/33. 

Sorry we didn't make it clear that a corresponding GitHub issue had to be created first.

**Comment by MarkusReynolds1989** [comment]
> @MarkusReynolds1989 Thanks a ton for working on this! There is one step that has been skipped here, and that is having the initial exercise design discussion. For new Concept Exercises, we want to have an upfront discussion. This both serves as a way for other maintainers to keep up to date with what's happening and give their insights on the exercise, besides it also serving as vital documentation. As there is so much going on with Concept Exercises, having an upfront discussion also helps prevent work having to be re-done.
> 
> @MarkusReynolds1989 Would you mind creating the corresponding GitHub issue so we can discuss the exercise there? There already is an example issue #33.
> 
> Sorry we didn't make it clear that a corresponding GitHub issue had to be created first.

Thanks for the feedback, this is the first serious project I've contributed too so I'm very much a newbie.

**Comment by ErikSchierboom** [comment]
> Thanks for the feedback, this is the first serious project I've contributed too so I'm very much a newbie.

Oh don't take this as any criticism of any sorts! It's fantastic that you want to help and it is easy to miss these things when jumping in. No worries! 🌷

File: languages/csharp/concept-exercises/pattern-matching/.docs/introduction.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+﻿# Introduction
+
+Pattern matching in C# is very useful and recently
+has taken a turn to look more like other languages.
```
[the remark this answers is not in the document]
Comment (ErikSchierboom) [reply languages/csharp/concept-exercises/pattern-matching/.docs/introduction.md]: I don't think we do. We should focus on C# only.

File: languages/csharp/concept-exercises/pattern-matching/PatternMatching.csproj [hunk 2]
```diff
@@ -0,0 +1,22 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>netcoreapp3.0</TargetFramework>
+    <RootNamespace>pattern_matching</RootNamespace>
```
Comment (ErikSchierboom) [inline languages/csharp/concept-exercises/pattern-matching/PatternMatching.csproj]: This can be removed.

File: languages/csharp/concept-exercises/pattern-matching/PatternMatching.csproj [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  </PropertyGroup>
+
+  <ItemGroup>
+      <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.4.0" />
+      <PackageReference Include="xunit" Version="2.4.1" />
+      <PackageReference Include="xunit.runner.visualstudio" Version="2.4.1" />
+  </ItemGroup>
+
+  <ItemGroup>
```
Comment (ErikSchierboom) [inline languages/csharp/concept-exercises/pattern-matching/PatternMatching.csproj]: This section should be removed.

File: languages/csharp/concept-exercises/pattern-matching/PatternMatching.cs [hunk 4]
```diff
@@ -0,0 +1,10 @@
+﻿using System;
+
+//See the Shapes file for the types you'll need to support.
```
Comment (ErikSchierboom) [inline languages/csharp/concept-exercises/pattern-matching/PatternMatching.cs]: I'd prefer having the types in the same file. Makes for an easier review of the solution, as well as the fact that the in-browser coding doesn't support multiple files at the moment.

File: languages/csharp/concept-exercises/pattern-matching/.docs/introduction.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
+﻿# Introduction
+
+Pattern matching in C# is very useful because
```
Comment (ErikSchierboom) [inline languages/csharp/concept-exercises/pattern-matching/.docs/introduction.md]: I feel that the introduction is slightly verbose and perhaps could be condensed a bit. Consider the following:

> The use cases for pattern matching are innumerable,  
but the use-case that comes to mind first is for
building compilers and parsers.

I don't mind having the compilers reference, but it might be possibly to rephrase and shorten things a bit.

File: languages/csharp/concept-exercises/pattern-matching/.docs/hints.md [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+the object. The input is given for you in the test
+so all you need to do is complete the math based upon
+the objects classes.
+
+### 4. Check for edge cases
+
+- Make sure you check that the method is actually
+getting a value
+- You can do this several ways [Nullable Types](https://stackoverflow.com/questions/271588/passing-null-arguments-to-c-sharp-methods)
+- [Optional Arguments](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/named-and-optional-arguments)
```
Comment (ErikSchierboom) [inline languages/csharp/concept-exercises/pattern-matching/.docs/hints.md]: Is it necessary for this exercise to use optional arguments? If not, I'd prefer removing them as they make for an additional requirement. In general, we want our exercises to have the bare minimal number of prerequisites.

File: languages/csharp/concept-exercises/pattern-matching/.docs/after.md [hunk 7]
```diff
@@ -0,0 +1,3 @@
+# After
+
+- [Tutorial on pattern matching](https://docs.microsoft.com/en-us/dotnet/csharp/pattern-matching)
```
Comment (ErikSchierboom) [inline languages/csharp/concept-exercises/pattern-matching/.docs/after.md]: This document is also referred to in the `hints.md` file and does not really add much in terms of in-depth explanation. So I think I this can be removed, unless you found another document.

File: languages/csharp/concept-exercises/pattern-matching/.docs/hints.md [hunk 8]
```diff
@@ -0,0 +1,28 @@
+﻿# Hints
+
+### 1. Check the type
```
Comment (ErikSchierboom) [inline languages/csharp/concept-exercises/pattern-matching/.docs/hints.md]: I'll skip reviewing this file due to the exercise possibly changing due to a comment I made elsewhere.

File: languages/csharp/concept-exercises/pattern-matching/Example.cs [hunk 9]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
+
+public static class Example
+{
+    public static double Perimeter(object shape = null)
+    {
+        var perimeter = shape switch
+        {
+            Square s => s.Side * 4,
+            Circle c => (2 * Math.PI) * c.Radius,
```
[the remark this answers is not in the document]
Comment (ErikSchierboom) [reply languages/csharp/concept-exercises/pattern-matching/Example.cs]: Another great resource to take inspiration from is: https://docs.microsoft.com/en-us/dotnet/csharp/tutorials/pattern-matching

File: languages/csharp/concept-exercises/pattern-matching/.docs/instructions.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
+﻿# Instructions
```
Comment (ErikSchierboom) [inline languages/csharp/concept-exercises/pattern-matching/.docs/instructions.md]: Currently, the instructions spell out what the student is expected to do, even with some very concrete examples. However, we want our exercises to designed such that students will be guided to using the correct approach, but not by explicitly stating what they must do. In other words, when solving the exercise, we should design it that the students are highly likely to solve the issue in the way that teaches what Concepts we are trying to teach.

This can be quite tricky to get right, and is probably the most difficulty part of designing an exercise. In general, I'd like this exercise to focus just on the modern, C# 8 pattern matching as described [here](https://docs.microsoft.com/en-us/dotnet/csharp/pattern-matching). This means teaching:

1. The `is` type pattern expression.
1. The pattern matching `switch` statement.
1. The `when` clause in a `switch` statement.
1. The `var` expression in a `case` expression.
1. `switch` expressions

I think this is really the type of discussion we should be having in the GH issue, so I'll also add my comments there.

**Comment by ErikSchierboom** [comment]
@MarkusReynolds1989 I've added my thoughts to the corresponding GH issue: https://github.com/exercism/v3/issues/546#issuecomment-583454293 Let's have the other maintainers chime in there too, so that we get a consensus on what this exercise should be teaching.

Do note that I really appreciate that you want to build this exercise! In fact, I'm really excited about this exercise as I love the pattern matching functionality in C# 7+
